### PR TITLE
Fix to handle DRM key Session for different and same keys

### PR DESCRIPTION
**Pull request by RahulRaas**
This is the fix to handle multiple DRM key sessions for Widevine contents (In widevine contents, the key will be different for Audio and Video)

Successfully tested with Video encrypted Stream, Audio encrypted Stream and both ( Video&Audio) encrypted Stream with different keys.

Working Branches :
WPEWebKit : staging-emev3-wip
OpenCDM : wpe
OpenCDMi : opencdm

### CODE REVIEW

**Comment by charlie-ht** [comment]
@RahulRaas Another thing, can you publicly disclose what test you run for this? If possible it would be nice to use as a test if other developers need to modify these parts of the code.

**Comment by RahulRaas** [comment]
@charlie-ht , I didn't observed the GStreamer-CRITICAL error message in my testing. I will try to reproduce the issue here. Can you please share me the stream that you used for testing.

Regarding testing, I have created 3 encrypted stream using Shaka packager.
1) Video encrypted 
2) Audio encrypted
3) Video and Audio encrypted with different keys
I am playing these stream using EME_Test_App_V3.html app.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp [hunk 1]
```diff
@@ -363,4 +365,19 @@ static void webKitMediaCommonEncryptionDecryptDefaultReleaseCipher(WebKitMediaCo
 {
 }
 
+static gchar* getContentType(GstElement* element)
+{
+    gchar* contentType = nullptr;
+
+    GstPad* sinkPad = gst_element_get_static_pad(element, "sink");
+    GstCaps* elementCaps = gst_pad_get_current_caps(sinkPad);
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: You need to do the same smart pointer changes as for the other twin function.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp [hunk 2]
```diff
@@ -363,4 +365,19 @@ static void webKitMediaCommonEncryptionDecryptDefaultReleaseCipher(WebKitMediaCo
 {
 }
 
+static gchar* getContentType(GstElement* element)
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: Let's make this a String

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp [hunk 3]
```diff
@@ -363,4 +365,19 @@ static void webKitMediaCommonEncryptionDecryptDefaultReleaseCipher(WebKitMediaCo
...
+{
+    gchar* contentType = nullptr;
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: String

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 4]
```diff
@@ -89,24 +90,32 @@ static void webKitMediaOpenCDMDecryptorFinalize(GObject* object)
 
 static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEncryptionDecrypt* self, GstEvent* event)
 {
+    bool ret = false;
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: ret -> returnValue

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 5]
```diff
@@ -89,24 +90,32 @@ static void webKitMediaOpenCDMDecryptorFinalize(GObject* object)
...
     const GstStructure* structure = gst_event_get_structure(event);
     if (!gst_structure_has_name(structure, "drm-session"))
-        return false;
+        return ret;
 
     GUniqueOutPtr<char> temporarySession;
-    gst_structure_get(structure, "session", G_TYPE_STRING, &temporarySession.outPtr(), nullptr);
+    GUniqueOutPtr<char> tempContentType;
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Do not shorten variable names, style requires full words...

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 6]
```diff
@@ -89,24 +90,32 @@ static void webKitMediaOpenCDMDecryptorFinalize(GObject* object)
...
+    gst_structure_get(structure, "session", G_TYPE_STRING, &temporarySession.outPtr(),
+        "contentType", G_TYPE_STRING, &tempContentType.outPtr(), nullptr);
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Make it a single line

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 7]
```diff
@@ -89,24 +90,32 @@ static void webKitMediaOpenCDMDecryptorFinalize(GObject* object)
...
-
-    if (priv->m_session != temporarySession.get() ) {
-        priv->m_session = temporarySession.get();
-        priv->m_openCdm = std::make_unique<media::OpenCdm>();
-        priv->m_openCdm->SelectSession(priv->m_session.utf8().data());
-        GST_DEBUG_OBJECT(self, "selected session %s", priv->m_session.utf8().data());
+    ASSERT(tempContentType);
+
+    if (priv->m_session != temporarySession.get()) {
+        /* Comparing decrypt plugin contentType with initiator contentType */
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: // .

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 8]
```diff
@@ -89,24 +90,32 @@ static void webKitMediaOpenCDMDecryptorFinalize(GObject* object)
...
+        if (!(g_ascii_strncasecmp(getContentType(GST_ELEMENT(self)), tempContentType.get(), CONTENT_TYPE_SIZE))) {
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: What's the purpose of comparing only 5 characters? This seems wrong.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 9]
```diff
@@ -201,4 +210,20 @@ static gboolean webKitMediaOpenCDMDecryptorDecrypt(WebKitMediaCommonEncryptionDe
     return returnValue;
 }
 
+static gchar* getContentType(GstElement* element)
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: This function, unless I missed something is the same for the common class. Either make it public in the common decryptor header or move it to the GStreamerEMEUtilities file. I think the utilities file is better.

And it needs a more descriptive name is needed like getOriginalMediaTypeFromGstElement.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 10]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
                 GST_WARNING("cannot map %s protection data", eventKeySystemId);
                 break;
             }
+#if USE(OPENCDM)
+            /* Getting source (qtdemux) content type and initData */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 11]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
+            String contentType = getContentType(GST_MESSAGE_SRC(message));
+            uint8_t* initData = (uint8_t*)malloc(mapInfo.size);
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 12]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
+            memcpy(initData, reinterpret_cast<uint8_t*>(mapInfo.data), mapInfo.size);
+
+            /* Compare previously saved initData with received initData */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 13]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
+            uint8_t initDataIt = 0;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 14]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
+            if (checkInitData(initData, initDataIt)) {
+                m_initDataInfo[initDataIt].contentType.append(contentType);
+                m_handledProtectionEvents.add(GST_EVENT_SEQNUM(event.get()));
+                GST_TRACE("m_handledProtectionEvents count %x", GST_EVENT_SEQNUM(event.get()));
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 15]
```diff
@@ -430,6 +430,23 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
...
+                return false;
+            }
+            /* Creating initData and content type map */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 16]
```diff
@@ -1361,9 +1378,14 @@ void MediaPlayerPrivateGStreamerBase::attemptToDecryptWithLocalInstance()
     if (is<CDMInstanceOpenCDM>(*m_cdmInstance)) {
         GST_DEBUG("handling OpenCDM %s keys", m_cdmInstance->keySystem().utf8().data());
         auto& cdmInstanceOpenCDM = downcast<CDMInstanceOpenCDM>(*m_cdmInstance);
-        String sessionId = cdmInstanceOpenCDM.getCurrentSessionId();
-        ASSERT(!sessionId.isEmpty());
-        dispatchDecryptionSession(sessionId);
+        uint8_t* initData;
+        String sessionId;
+        bool ret = cdmInstanceOpenCDM.getCurrentSessionInfo(sessionId, initData);
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 17]
```diff
@@ -1361,9 +1378,14 @@ void MediaPlayerPrivateGStreamerBase::attemptToDecryptWithLocalInstance()
...
+        if (ret) {
+            m_sessionInfoMap.add(sessionId, initData);
+            validateDecryptionSession(sessionId, initData);
+        } else
+            GST_WARNING("Not able to retrieve current session info");
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 18]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
 
-void MediaPlayerPrivateGStreamerBase::dispatchDecryptionSession(const String& sessionId)
+void MediaPlayerPrivateGStreamerBase::dispatchDecryptionSession(const String& sessionId, const String& contentType)
 {
     bool eventHandled = gst_element_send_event(m_pipeline.get(), gst_event_new_custom(GST_EVENT_CUSTOM_DOWNSTREAM_OOB,
-        gst_structure_new("drm-session", "session", G_TYPE_STRING, sessionId.utf8().data(), nullptr)));
-    m_needToResendCredentials = m_handledProtectionEvents.size() > 0;
-    GST_TRACE("emitted decryption session %s on pipeline, event handled %s, need to resend credentials %s", sessionId.utf8().data(), boolForPrinting(eventHandled), boolForPrinting(m_needToResendCredentials));
+        gst_structure_new("drm-session", "session", G_TYPE_STRING, sessionId.utf8().data(), "contentType",
+        G_TYPE_STRING, contentType.utf8().data(), nullptr)));
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 19]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    GST_TRACE("emitted decryption session %s on pipeline, event handled %s", sessionId.utf8().data(), boolForPrinting(eventHandled));
+}
+
+#if USE(OPENCDM)
+void MediaPlayerPrivateGStreamerBase::validateDecryptionSession(const String& sessionId, const uint8_t* initData)
+{
+    uint8_t initDataIt = 0;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 20]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+
+    if (!m_decryptorCount)
+        return;
+
+    /* Compare initData to get initiator content type */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 21]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    if (checkInitData(initData, initDataIt)) {
+        for (uint8_t contentTypeIt = 0; contentTypeIt < m_initDataInfo[initDataIt].contentType.size(); contentTypeIt++) {
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 22]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+            /* Check initiator content type with all decrypt plugin content type */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 23]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+            if (m_decryptorContentType.find(m_initDataInfo[initDataIt].contentType[contentTypeIt]) > 0) {
+                /*  If content type found then send the event */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 24]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+                dispatchDecryptionSession(sessionId, m_initDataInfo[initDataIt].contentType[contentTypeIt]);
+                m_initDataInfo[initDataIt].contentType.remove(contentTypeIt);
+            }
+        }
+    }
 }
 
-void MediaPlayerPrivateGStreamerBase::handleProtectionEvent(GstEvent* event)
+String MediaPlayerPrivateGStreamerBase::getContentType(const GstObject* object)
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 25]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+{
+    GstPad* sinkPad = gst_element_get_static_pad(GST_ELEMENT(object), "sink");
+    GstCaps* elementCaps = gst_pad_get_current_caps(sinkPad);
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 26]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    GstStructure* structure = gst_caps_get_structure(elementCaps, 0);
+    String contentType = (gchar*)gst_structure_get_name(structure);
+
+    gst_caps_unref(elementCaps);
+    gst_object_unref(sinkPad);
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 27]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    }
 }
 
-void MediaPlayerPrivateGStreamerBase::handleProtectionEvent(GstEvent* event)
+String MediaPlayerPrivateGStreamerBase::getContentType(const GstObject* object)
+{
+    GstPad* sinkPad = gst_element_get_static_pad(GST_ELEMENT(object), "sink");
+    GstCaps* elementCaps = gst_pad_get_current_caps(sinkPad);
+    GstStructure* structure = gst_caps_get_structure(elementCaps, 0);
+    String contentType = (gchar*)gst_structure_get_name(structure);
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 28]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    gst_object_unref(sinkPad);
+
+    return contentType;
+}
+
+bool MediaPlayerPrivateGStreamerBase::checkInitData(const uint8_t* initData, uint8_t& index)
+{
+    bool ret = false;
+
+    for (uint8_t iterator = 0; iterator < m_initDataInfo.size(); iterator++) {
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 29]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+        if (!(memcmp(initData, m_initDataInfo[iterator].initData, m_initDataInfo[iterator].initDataSize))) {
+            ret =true;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 30]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+        }
+    }
+    return ret;
+}
+
+void MediaPlayerPrivateGStreamerBase::createInitDataMap(uint8_t*& initData, uint8_t initDataSize, String& contentType)
+{
+    InitDataInfo initDataInfo;
+    initDataInfo.initData = initData;
+    initDataInfo.initDataSize =  initDataSize;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 31]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
+    initDataInfo.contentType.append(contentType);
+    m_initDataInfo.append(initDataInfo);
+}
+#endif
+
+void MediaPlayerPrivateGStreamerBase::handleProtectionEvent(GstEvent* event, gchar* contentType)
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 32]
```diff
@@ -1376,23 +1398,87 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
...
 {
     if (m_handledProtectionEvents.contains(GST_EVENT_SEQNUM(event))) {
         GST_DEBUG("event %u already handled", GST_EVENT_SEQNUM(event));
         m_handledProtectionEvents.remove(GST_EVENT_SEQNUM(event));
-        if (m_needToResendCredentials) {
-            GST_DEBUG("resending credentials");
-            attemptToDecryptWithLocalInstance();
+#if USE(OPENCDM)
+        if (!m_sessionInfoMap.isEmpty()) {
+            /* Send GST_EVENT_CUSTOM_DOWNSTREAM_OOB event with all saved SessionId */
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h [hunk 33]
```diff
@@ -40,10 +40,25 @@
 #include "TextureMapperPlatformLayerProxyProvider.h"
 #endif
 
+#if USE(OPENCDM)
+/* Now we are only handling Audio and Video content type, so defining CONTENT_TYPE_SIZE as 5.
+   We will modify it once we have subtitle */
+#define CONTENT_TYPE_SIZE 5
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h [hunk 34]
```diff
@@ -40,10 +40,25 @@
...
+
+struct _InitDataInfo {
+    uint8_t* initData;
+    uint8_t initDataSize;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h [hunk 35]
```diff
@@ -40,10 +40,25 @@
...
+    Vector<String> contentType;
+};
+#endif
+
 typedef struct _GstStreamVolume GstStreamVolume;
 typedef struct _GstVideoInfo GstVideoInfo;
 typedef struct _GstGLContext GstGLContext;
 typedef struct _GstGLDisplay GstGLDisplay;
+#if USE(OPENCDM)
+typedef struct _InitDataInfo InitDataInfo;
```
[the remark this answers is not in the document]
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.h]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 4 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 5 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 6 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 7 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp (hunk 1 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp (hunk 3 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: Resolved

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp (hunk 2 above)
Comment (RahulRaas) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp]: Resolved

**Pull request closed without merging**